### PR TITLE
Add User related APIs

### PR DESCRIPTION
**Pull request by Noah02148**
Revision2:
Address comments.

Use AfterInsert() to update followers/followings count and itemSaves/itemLikes count

---------------

Main changes:

1. Rename Follower entity to UserRelation entity,  add followers and followings as UserRelation instances in to User entity.

2. Update itemSaves, itemLikes APIs.

Testing:
GET {{HOST}}/v1/users/1/followings

```
{
    "data": {
        "id": 1,
        "openId": null,
        "username": "fangzhou",
        "mobilePrefix": "+86",
        "mobile": "15828332840",
        "role": null,
        "createdAt": "2021-05-14T11:22:55.440Z",
        "updatedAt": "2021-05-14T11:22:55.440Z",
        "followings": [
            {
                "id": "d8d2d53f-e775-4ddf-b751-624a8fa46fba",
                "createdAt": "2021-05-14T14:30:31.802Z",
                "updatedAt": "2021-05-14T14:30:31.802Z",
                "followee": {
                    "id": 2,
                    "openId": null,
                    "username": "concepts",
                    "mobilePrefix": "+86",
                    "mobile": "15828332833",
                    "role": null,
                    "createdAt": "2021-05-14T14:24:52.166Z",
                    "updatedAt": "2021-05-14T14:24:52.166Z",
                    "followingsCount": 0,
                    "followersCount": 1
                }
            },
            {
                "id": "ce1932ee-aefb-4168-97d4-c04a0fe538cf",
                "createdAt": "2021-05-14T14:40:41.689Z",
                "updatedAt": "2021-05-14T14:40:41.689Z",
                "followee": {
                    "id": 3,
                    "openId": null,
                    "username": "bodega",
                    "mobilePrefix": "+86",
                    "mobile": "15828402840",
                    "role": null,
                    "createdAt": "2021-05-14T14:39:45.506Z",
                    "updatedAt": "2021-05-14T14:39:45.506Z",
                    "followingsCount": 1,
                    "followersCount": 1
                }
            }
        ],
        "followingsCount": 2,
        "followersCount": 1
    }
}

```

GET {{HOST}}/v1/users/1/followers

```

{
    "data": {
        "id": 1,
        "openId": null,
        "username": "fangzhou",
        "mobilePrefix": "+86",
        "mobile": "15828332840",
        "role": null,
        "createdAt": "2021-05-14T11:22:55.440Z",
        "updatedAt": "2021-05-14T11:22:55.440Z",
        "followers": [
            {
                "id": "2bc154ad-0071-4745-8bec-52cf7cd93ed8",
                "createdAt": "2021-05-14T15:17:06.742Z",
                "updatedAt": "2021-05-14T15:17:06.742Z",
                "follower": {
                    "id": 3,
                    "openId": null,
                    "username": "bodega",
                    "mobilePrefix": "+86",
                    "mobile": "15828402840",
                    "role": null,
                    "createdAt": "2021-05-14T14:39:45.506Z",
                    "updatedAt": "2021-05-14T14:39:45.506Z",
                    "followingsCount": 1,
                    "followersCount": 1
                }
            }
        ],
        "followingsCount": 2,
        "followersCount": 1
    }
}
```

GET {{HOST}}/v1/users/1/likedItems
```
{
    "data": {
        "id": 1,
        "openId": null,
        "username": "fangzhou",
        "mobilePrefix": "+86",
        "mobile": "15828332840",
        "role": null,
        "createdAt": "2021-05-14T11:22:55.440Z",
        "updatedAt": "2021-05-14T11:22:55.440Z",
        "itemLikes": [
            {
                "id": "5785de05-5fa1-4b1f-a5da-61ed71b35962",
                "createdAt": "2021-05-14T13:12:48.330Z",
                "updatedAt": "2021-05-14T13:12:48.330Z",
                "item": {
                    "id": "2492aa22-1d61-49c7-a9a6-922940e8ddd3",
                    "name": "dandyhobo items 1",
                    "price": 100,
                    "condition": "like new",
                    "color": "black",
                    "size": "M",
                    "imageUrls": "fake_url",
                    "factoryDate": "1990s",
                    "description": "steal",
                    "status": null,
                    "createdAt": "2021-05-14T12:52:08.965Z",
                    "updatedAt": "2021-05-14T12:52:08.965Z",
                    "itemSavesCount": 1,
                    "itemLikesCount": 1
                }
            }
        ],
        "followingsCount": 2,
        "followersCount": 1
    }
}

```

GET {{HOST}}/v1/users/1/savedItems

```
{
    "data": {
        "id": 1,
        "openId": null,
        "username": "fangzhou",
        "mobilePrefix": "+86",
        "mobile": "15828332840",
        "shopId": null,
        "role": null,
        "createdAt": "2021-05-14T11:22:55.440Z",
        "updatedAt": "2021-05-14T11:22:55.440Z",
        "itemSaves": [
            {
                "id": "c96ecf90-5868-4b08-9033-eb1e139b7422",
                "createdAt": "2021-05-14T12:50:31.669Z",
                "updatedAt": "2021-05-14T12:50:31.669Z",
                "item": {
                    "id": "06439cc5-33de-49df-ab3e-1e8dc1930495",
                    "name": "dandyhobo items 1",
                    "price": 100,
                    "condition": "like new",
                    "color": "black",
                    "size": "M",
                    "imageUrls": "fake_url",
                    "factoryDate": "1990s",
                    "description": "steal",
                    "status": null,
                    "createdAt": "2021-05-14T11:32:34.290Z",
                    "updatedAt": "2021-05-14T11:32:34.290Z"
                }
            },
            {
                "id": "71f209d8-5023-4cc9-81a9-3589fb9719bf",
                "createdAt": "2021-05-14T12:52:26.375Z",
                "updatedAt": "2021-05-14T12:52:26.375Z",
                "item": {
                    "id": "2492aa22-1d61-49c7-a9a6-922940e8ddd3",
                    "name": "dandyhobo items 1",
                    "price": 100,
                    "condition": "like new",
                    "color": "black",
                    "size": "M",
                    "imageUrls": "fake_url",
                    "factoryDate": "1990s",
                    "description": "steal",
                    "status": null,
                    "createdAt": "2021-05-14T12:52:08.965Z",
                    "updatedAt": "2021-05-14T12:52:08.965Z"
                }
            }
        ],
        "followingsCount": 2,
        "followersCount": 1
    }
}

```

### CODE REVIEW

File: src/controller/itemLike.ts [hunk 1]
```diff
@@ -18,15 +19,20 @@ export class ItemLikeController {
...
+    const user = await getRepository(Users).findOne({id: userId});
+    if (!user) {
+      throw new ResourceNotFoundError("User is not found.");
+    }
 
-    const itemLike = await itemLikeRepo.findOne({likedBy: userId, item: item});
+    const itemLikeRepo = getRepository(ItemLikes);
+    const itemLike = await itemLikeRepo.findOne({user: user, item: item});
     if (itemLike) {
       logger.warning("ItemLike already exists.");
```
Comment (bigdaddyrus) [inline src/controller/itemLike.ts:30]: if exist 直接return itemlike

File: src/controller/itemSave.ts [hunk 2]
```diff
@@ -19,16 +20,20 @@ export class ItemSaveController {
...
+    const user = await getRepository(Users).findOne({id: userId});
+    if (!user) {
+      throw new ResourceNotFoundError("User is not found.");
+    }
 
-    const itemSave = await itemSaveRepo.findOne({savedBy: userId, item: item});
+    const itemSaveRepo = getRepository(ItemSaves);
+    const itemSave = await itemSaveRepo.findOne({user: user, item: item});
     if (itemSave) {
       logger.warning("ItemSave already exists.");
```
Comment (bigdaddyrus) [inline src/controller/itemSave.ts:31]: 同上

File: src/controller/userRelation.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    const followee = await userRepo.findOne({id: followeeId});
+    if (!followee) {
+      throw new ResourceNotFoundError("Followee is not found.");
+    }
+
+    const userRelationRepo = getRepository(UserRelations);
+    const userRelationEntry = await userRelationRepo.findOne({ follower: follower, followee: followee });
+    if (userRelationEntry) {
+      logger.warning("UserRelationEntry already exists.");
```
Comment (bigdaddyrus) [inline src/controller/userRelation.ts:31]: 同上

File: src/controller/userRelation.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import { HandleError } from "../decorator/errorDecorator";
+import { ResourceNotFoundError } from "../error/notfoundError";
+import { Users } from "../entities/Users";
+import { UserRelations } from "../entities/UserRelations";
+import { logger } from "../logging/logger";
+import { RequestValidator } from "../validator/requestValidator";
+
+export class UserRelationController {
+
+  @HandleError("createFollow")
```
Comment (bigdaddyrus) [inline src/controller/userRelation.ts:12]: unfollow?

File: src/entities/Users.ts [hunk 5]
```diff
@@ -49,6 +53,24 @@ export class Users extends BaseEntity {
...
+  @OneToMany(() => ItemLikes, (itemLikes) => itemLikes.user, { cascade: true })
+  itemLikes: ItemLikes[];
+
+  @OneToMany(type => UserRelations, userRelations => userRelations.follower)
+  followings: UserRelations[];
+
+  @OneToMany(type => UserRelations, userRelations => userRelations.followee)
+  followers: UserRelations[];
+
+  @RelationCount((user: Users) => user.followings)
```
Comment (bigdaddyrus) [inline src/entities/Users.ts:68]: 这个貌似deprecated 了： https://stackoverflow.com/questions/58418569/typeorm-relationcount-alternative

File: src/v1router.ts [hunk 6]
```diff
@@ -18,12 +18,16 @@ v1router.use(authMiddleWare);
...
 v1router.post("/items/:id/like", ItemLikeController.createItemLike);
 v1router.post("/items/:id/save", ItemSaveController.createItemSave);
-v1router.post("/followers/:id/follow", FollowerController.createFollow);
+v1router.post("/users/:id/follow", UserRelationController.createFollow);
 v1router.post("/addresses", AddressController.createAddress);
 
 // GET
 v1router.get("/items/:id", ItemController.getItem);
 v1router.get("/addresses", AddressController.getAddresses);
+v1router.get("/users/:id/savedItems", ItemSaveController.getUserSavedItems);
```
Comment (bigdaddyrus) [inline src/v1router.ts:27]: 感觉group by resource instead of http action verb 比较好读

Comment (Noah02148) [reply src/v1router.ts:27]: 可

File: src/controller/userRelation.ts (hunk 4 above)
Comment (Noah02148) [reply src/controller/userRelation.ts:12]: will add in next pr

**Pull request merged**